scripts/build/check_init_priorities.py

```diff
@@ -0,0 +1,319 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+Checks the initialization priorities
+
+This script parses the object files in the specified build directory, creates a
+list of known devices and their effective initialization priorities and
+compares that with the device dependencies inferred from the devicetree
+hierarchy.
+
+This can be used to detect devices that are initialized in the incorrect order,
+but also devices that are initialized at the same priority but depends on each
+other, which can potentially break if the linking order is changed.
+"""
+
+import argparse
+import logging
+import os
+import pathlib
+import pickle
+import sys
+
+from elftools.elf.elffile import ELFFile
+from elftools.elf.relocation import RelocationSection
+from elftools.elf.sections import SymbolTableSection
+
+# This is needed to load edt.pickle files.
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..",
+                                "dts", "python-devicetree", "src"))
+from devicetree import edtlib  # pylint: disable=unused-import
+
+# Prefix used for relocation sections containing initialization data, as in
+# sequence of "struct init_entry".
+_INIT_SECTION_PREFIX = (".rel.z_init_", ".rela.z_init_")
+
+# Prefix used for "struct device" reference initialized based on devicetree
+# entries with a known ordinal.
+_DEVICE_ORD_PREFIX = "__device_dts_ord_"
+
+# File name suffix for object files to be scanned.
+_OBJ_FILE_SUFFIX = ".c.obj"
+
+# Defined init level in order of priority.
+_DEVICE_INIT_LEVELS = ["EARLY", "PRE_KERNEL_1", "PRE_KERNEL_2", "POST_KERNEL",
+                      "APPLICATION", "SMP"]
+
+# Build project settings file, used to detect and skip sub-builds.
+_SETTINGS_FILE = "zephyr_settings.txt"
```

Comment:
Yeah, not ideal, so I need something to identify a subbuild, initially used `build.ninja` but then it would not work with Makefile since that generates a Makefile for every subdirectory, was left with `zephyr_modules.txt` and `zephyr_settings.txt`. I could look for `CMakeCache.txt`? Or maybe add an explicit marker in the build directory, say adding a `touch ${PROJECT_BINARY_BUILD}/.zephyr_build` somewhere in the build system and look for that instead.

If we get to the point of enabling this in CI it would fail and catch it if module system gets reworked, but it'd be cool to not depend on that in the first place. Can you think of any better strategy?

Reply:
just to confirm my understanding.

You want to avoid processing sub-builds build directories, such as a TF-M build and similar, for performance reasons ?

Such sub-builds can be
1.  additional Zephyr builds
2. other CMake based builds, like [TF-M](https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/trusted-firmware-m/)
3. a Makefile based project, like [external_lib](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/application_development/external_lib).
4. Other / custom type projects

If this is the case, the `CMakeCache.txt` approach is safest cause that will catch 1 & 2, whereas current approach will only catch 1, and bear the built-in risk mentioned earlier.
3. might be present in a build but often represents integration to legacy systems, but we can't do much for those.
4. I wouldn't worry too much.

Reply:
> just to confirm my understanding.
> 
> You want to avoid processing sub-builds build directories, such as a TF-M build and similar, for performance reasons ?

Yes but not for performance reasons, the script goes through all the object files looking for defined device structs and mapping their priority, if there are nested (Zephyr) projects they can potentially redefine the same device ordinal at different priority (or confuse the script in other ways). The script detects the double define (the two "device already defined" exceptions) but then does not know what to do with it.

> If this is the case, the `CMakeCache.txt` approach is safest cause that will catch 1 & 2, whereas current approach will only catch 1, and bear the built-in risk mentioned earlier. 3. might be present in a build but often represents integration to legacy systems, but we can't do much for those. 4. I wouldn't worry too much.

Sounds good, swapped for `CMakeCache.txt`, changed few constant names and comments.